climada/util/coordinates.py

```diff
@@ -76,6 +78,273 @@
+    * util.lines_polys_handler.point_exposure_from_lines()
+    """
+
+    if not isinstance(gdf_lines, gpd.GeoDataFrame):
+        gdf_lines = gpd.read_file(gdf_lines)
+        
+    gdf_points = gdf_lines.copy()
+    gdf_points['length'] = np.ones(len(gdf_points))*point_dist
+    gdf_points['length_full'] = compute_geodesic_lengths(gdf_points)
+
+    # split line lengths into relative fractions acc to point_dist (e.g. 0, 0.5, 1)
+    gdf_points['distance_vector'] = gdf_points.apply(
+        lambda row: np.linspace(0, 1, num=int(np.ceil(row.length_full/
+                                                      row.length)+1)), axis=1)
+
+    # create MultiPoints along the line for every position in distance_vector
+    gdf_points['geometry'] = gdf_points.apply(
+        lambda row: MultiPoint(
+            [row.geometry.interpolate(dist, normalized=True) 
+             for dist in row.distance_vector]), axis=1)
+    
+    # expand gdf from MultiPoint entries to single Points per row
+    return gdf_points.explode().drop(['distance_vector', 'length_full'], axis=1)
+
+def interpolate_polygons(gdf_poly, area_point):
+    """For a GeoDataFrame with polygons, get equally distributed lat/lon pairs
+    throughout the geometries, at a user-specified area distance
+    
+    Parameters
+    ----------
+    gdf_poly : gpd.GeoDataFrame
+        with polygons to be interpolated
+    area_point : float
+        area in m2 which one point should represent
+    
+    Returns
+    -------
+    (gpd.GeoDataFrame) with individual Point per row, retaining all other column infos
+        belonging to its corresponding polygon
+    
+    See also
+    --------
+    * util.coordinates.compute_geodesic_lengths()
+    * util.lines_polys_handler.point_exposure_from_polygons()
+    """
+    
+    metre_dist = math.sqrt(area_point)
+    gdf_points = gdf_poly.copy()
+    
+    gdf_points['degree_dist'] = gdf_poly.apply(lambda row: metres_to_degrees(
+        row.geometry.representative_point().x, 
+        row.geometry.representative_point().y,
+        metre_dist), axis=1)
+    
+    # get params to make an even grid with desired resolution 
+    # over bounding boxes of polygons:
+    gdf_points[['height','width','trans']] = gdf_points.apply(
+        lambda row: pts_to_raster_meta(row.geometry.bounds, 
+                                       (row.degree_dist, -row.degree_dist)), 
+        axis=1).tolist()
+    
+    # make grid --> lon / lat for each polygon-bbox
+    gdf_points[['lon','lat']] = pd.DataFrame(gdf_points.apply(lambda row: 
+        raster_to_meshgrid(row.trans, row.width, row.height), axis=1).tolist())
+    
+    for axis in ['lon', 'lat']:
+        gdf_points[axis] = gdf_points.apply(lambda row: row[axis].flatten(), 
+                                            axis=1)
+    # filter only centroids in actual polygons
+    for i, polygon in enumerate(gdf_points.geometry):
+        in_geom = coord_on_land(lat=gdf_points['lat'].iloc[i], 
```

Comment:
Why implicitly assume points on land? The util function to transform from polygons to points should not assume that. There are several use cases where points on the ocean are relevant.

Reply:
That's not what it assumes. it does as said in the comment -  # filter only centroids in actual polygons
The function `coord_on_land` simply checks whether the points generated from the even grid lie within the desired polygon, and excludes the others. in short, it's just a wrapper for an sjoin func

Reply:
This is confusing to me. Why use `coord_on_land` then? What happens if the coords or the centroids are in the water and the polygon also? 

Why not use sjoin directly then?

Reply:
Thanks, I looked again at the code for `coord_on_land`, my bad. It can indeed be misused for filtering points in polygons.

Reply:
I checked the `coord_on_land` and it's basically a re-naming of the shapely-native function `vectorized.contains()` which is exactly what is needed here to see which points fall into the polygon. I use the latter directly now in commit e892d39a